Investigate replacing double_conversion library with fast_float

Constellation opened the pull request:
#### 178e0726ac6b62ca709507c98f17dd278dce5d8a
<pre>
Investigate replacing double_conversion library with fast_float
<a href="https://bugs.webkit.org/show_bug.cgi?id=221178">https://bugs.webkit.org/show_bug.cgi?id=221178</a>

Reviewed by Sam Weinig.

Parsing double numbers becomes 2x faster.

                                  ToT                     Patched

    json-parse-double      230.0022+-0.0838     ^    111.1297+-0.0983        ^ definitely 2.0697x faster

Apple OSS approval for fast_float is OSS-4890.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FastFloat.cpp: Added.
(WTF::parseDouble):
* Source/WTF/wtf/FastFloat.h: Added.
* Source/WTF/wtf/dtoa.h:
* Source/WTF/wtf/fast_float/LICENSE: Added.
* Source/WTF/wtf/fast_float/ascii_number.h: Added.
(fast_float::byteswap):
(fast_float::read_u64):
(fast_float::write_u64):
(fast_float::parse_eight_digits_unrolled):
* Source/WTF/wtf/fast_float/bigint.h: Added.
(fast_float::stackvec::stackvec):
(fast_float::bigint::bigint):
* Source/WTF/wtf/fast_float/decimal_to_binary.h: Added.
(fast_float::compute_product_approximation):
* Source/WTF/wtf/fast_float/digit_comparison.h: Added.
* Source/WTF/wtf/fast_float/fast_float.h: Added.
(fast_float::parse_options::parse_options):
* Source/WTF/wtf/fast_float/fast_table.h: Added.
* Source/WTF/wtf/fast_float/float_common.h: Added.
(fast_float::fastfloat_strncasecmp):
(fast_float::span::span):
(fast_float::value128::value128):
(fast_float::leading_zeroes):
(fast_float::emulu):
(fast_float::_umul128):
(fast_float::full_multiplication):
(fast_float::adjusted_mantissa::operator== const):
(fast_float::adjusted_mantissa::operator!= const):
(fast_float::binary_format&lt;double&gt;::mantissa_explicit_bits):
(fast_float::binary_format&lt;float&gt;::mantissa_explicit_bits):
(fast_float::binary_format&lt;double&gt;::max_exponent_round_to_even):
(fast_float::binary_format&lt;float&gt;::max_exponent_round_to_even):
(fast_float::binary_format&lt;double&gt;::min_exponent_round_to_even):
(fast_float::binary_format&lt;float&gt;::min_exponent_round_to_even):
(fast_float::binary_format&lt;double&gt;::minimum_exponent):
(fast_float::binary_format&lt;float&gt;::minimum_exponent):
(fast_float::binary_format&lt;double&gt;::infinite_power):
(fast_float::binary_format&lt;float&gt;::infinite_power):
(fast_float::binary_format&lt;double&gt;::sign_index):
(fast_float::binary_format&lt;float&gt;::sign_index):
(fast_float::binary_format&lt;double&gt;::min_exponent_fast_path):
(fast_float::binary_format&lt;float&gt;::min_exponent_fast_path):
(fast_float::binary_format&lt;double&gt;::max_exponent_fast_path):
(fast_float::binary_format&lt;float&gt;::max_exponent_fast_path):
(fast_float::binary_format&lt;double&gt;::max_mantissa_fast_path):
(fast_float::binary_format&lt;float&gt;::max_mantissa_fast_path):
(fast_float::binary_format&lt;double&gt;::exact_power_of_ten):
(fast_float::binary_format&lt;float&gt;::exact_power_of_ten):
(fast_float::binary_format&lt;double&gt;::largest_power_of_ten):
(fast_float::binary_format&lt;float&gt;::largest_power_of_ten):
(fast_float::binary_format&lt;double&gt;::smallest_power_of_ten):
(fast_float::binary_format&lt;float&gt;::smallest_power_of_ten):
(fast_float::binary_format&lt;double&gt;::max_digits):
(fast_float::binary_format&lt;float&gt;::max_digits):
(fast_float::binary_format&lt;float&gt;::exponent_mask):
(fast_float::binary_format&lt;double&gt;::exponent_mask):
(fast_float::binary_format&lt;float&gt;::mantissa_mask):
(fast_float::binary_format&lt;double&gt;::mantissa_mask):
(fast_float::binary_format&lt;float&gt;::hidden_bit_mask):
(fast_float::binary_format&lt;double&gt;::hidden_bit_mask):
(fast_float::to_float):
* Source/WTF/wtf/fast_float/parse_number.h: Added.
* Source/WTF/wtf/fast_float/simple_decimal_conversion.h: Added.
(fast_float::detail::trim):
(fast_float::detail::number_of_digits_decimal_left_shift):
(fast_float::detail::round):
(fast_float::detail::decimal_left_shift):
(fast_float::detail::decimal_right_shift):
(fast_float::compute_float):
(fast_float::parse_long_mantissa):

Canonical link: <a href="https://commits.webkit.org/257675@main">https://commits.webkit.org/257675@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/900ddd2bf14211a0b301934e33bc8fca4442741b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99095 "2053 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108489 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168731 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85647 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91599 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106432 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104820 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6691 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90269 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33706 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88513 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21614 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76564 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/89808 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2192 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23130 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/85635 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2084 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45566 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28842 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5281 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/7046 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42608 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/88494 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3500 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19806 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WTF/wtf/FastFloat.h:
```diff
@@ -0,0 +1,34 @@
+
+namespace WTF {
+
+WTF_EXPORT_PRIVATE double parseDouble(const LChar* string, size_t length, size_t& parsedLength);
```
I I think the way I would structured things would be to rename this file to ParseDouble.h and add all the ParseDouble header bits from dtoa.h here.

I would then name the file you have called FastFloat.cpp to ParseDoubleFastFloat.cpp.

And then, probably in a different change, rename dtoa.h to NumberToString.h and dtoa.cpp to NumberToStringDoubleConversion.cpp

weinig commented on Source/WTF/wtf/fast_float/ascii_number.h:
```diff
@@ -0,0 +1,231 @@
+
+#include <wtf/fast_float/float_common.h>
+
+namespace fast_float {
```
For whatever reason, we wrapped all the double_conversion namespaces in an our `namespace WTF`. Not sure it makes a difference given this is all only included in one cpp file, but maybe for mangling reasons it would be a good idea?

weinig reviewed (approved):
Thanks for following up on this!

Constellation committed:
Investigate replacing double_conversion library with fast_float
https://bugs.webkit.org/show_bug.cgi?id=221178

Reviewed by Sam Weinig.

Parsing double numbers becomes 2x faster.

                                  ToT                     Patched

    json-parse-double      230.0022+-0.0838     ^    111.1297+-0.0983        ^ definitely 2.0697x faster

Apple OSS approval for fast_float is OSS-4890.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/FastFloat.cpp: Added.
(WTF::parseDouble):
* Source/WTF/wtf/FastFloat.h: Added.
* Source/WTF/wtf/dtoa.h:
* Source/WTF/wtf/fast_float/LICENSE: Added.
* Source/WTF/wtf/fast_float/ascii_number.h: Added.
(fast_float::byteswap):
(fast_float::read_u64):
(fast_float::write_u64):
(fast_float::parse_eight_digits_unrolled):
* Source/WTF/wtf/fast_float/bigint.h: Added.
(fast_float::stackvec::stackvec):
(fast_float::bigint::bigint):
* Source/WTF/wtf/fast_float/decimal_to_binary.h: Added.
(fast_float::compute_product_approximation):
* Source/WTF/wtf/fast_float/digit_comparison.h: Added.
* Source/WTF/wtf/fast_float/fast_float.h: Added.
(fast_float::parse_options::parse_options):
* Source/WTF/wtf/fast_float/fast_table.h: Added.
* Source/WTF/wtf/fast_float/float_common.h: Added.
(fast_float::fastfloat_strncasecmp):
(fast_float::span::span):
(fast_float::value128::value128):
(fast_float::leading_zeroes):
(fast_float::emulu):
(fast_float::_umul128):
(fast_float::full_multiplication):
(fast_float::adjusted_mantissa::operator== const):
(fast_float::adjusted_mantissa::operator!= const):
(fast_float::binary_format<double>::mantissa_explicit_bits):
(fast_float::binary_format<float>::mantissa_explicit_bits):
(fast_float::binary_format<double>::max_exponent_round_to_even):
(fast_float::binary_format<float>::max_exponent_round_to_even):
(fast_float::binary_format<double>::min_exponent_round_to_even):
(fast_float::binary_format<float>::min_exponent_round_to_even):
(fast_float::binary_format<double>::minimum_exponent):
(fast_float::binary_format<float>::minimum_exponent):
(fast_float::binary_format<double>::infinite_power):
(fast_float::binary_format<float>::infinite_power):
(fast_float::binary_format<double>::sign_index):
(fast_float::binary_format<float>::sign_index):
(fast_float::binary_format<double>::min_exponent_fast_path):
(fast_float::binary_format<float>::min_exponent_fast_path):
(fast_float::binary_format<double>::max_exponent_fast_path):
(fast_float::binary_format<float>::max_exponent_fast_path):
(fast_float::binary_format<double>::max_mantissa_fast_path):
(fast_float::binary_format<float>::max_mantissa_fast_path):
(fast_float::binary_format<double>::exact_power_of_ten):
(fast_float::binary_format<float>::exact_power_of_ten):
(fast_float::binary_format<double>::largest_power_of_ten):
(fast_float::binary_format<float>::largest_power_of_ten):
(fast_float::binary_format<double>::smallest_power_of_ten):
(fast_float::binary_format<float>::smallest_power_of_ten):
(fast_float::binary_format<double>::max_digits):
(fast_float::binary_format<float>::max_digits):
(fast_float::binary_format<float>::exponent_mask):
(fast_float::binary_format<double>::exponent_mask):
(fast_float::binary_format<float>::mantissa_mask):
(fast_float::binary_format<double>::mantissa_mask):
(fast_float::binary_format<float>::hidden_bit_mask):
(fast_float::binary_format<double>::hidden_bit_mask):
(fast_float::to_float):
* Source/WTF/wtf/fast_float/parse_number.h: Added.
* Source/WTF/wtf/fast_float/simple_decimal_conversion.h: Added.
(fast_float::detail::trim):
(fast_float::detail::number_of_digits_decimal_left_shift):
(fast_float::detail::round):
(fast_float::detail::decimal_left_shift):
(fast_float::detail::decimal_right_shift):
(fast_float::compute_float):
(fast_float::parse_long_mantissa):

Canonical link: https://commits.webkit.org/257675@main
